cpp/src/Slice/Parser.cpp

```diff
@@ -2503,6 +2503,22 @@ Slice::ClassDef::createDataMember(const string& name, const TypePtr& type, bool
         return nullptr;
     }
 
+    // Check whether enclosing class has the same name.
+    if (name == this->name())
+    {
+        _unit->error("data member `" + name + "' cannot have the same name as its enclosing class");
+        _unit->note(this, "class `" + this->name() + "' is defined here");
+        return nullptr;
+    }
+
+    if (ciequals(name, this->name()))
+    {
+        _unit->error("data member `" + name + "' differs only in capitalization from its enclosing class named `"
+                     + this->name() + "'");
+        _unit->note(this, "class `" + this->name() + "' is defined here");
+        return nullptr;
+    }
+
```

Comment:
Maybe we can move this check to a common method

Reply:
I have in my next PR that unifies DataMember and ParamDecl. There's loads of logic that is shared between Struct, Class, Exception, and even somewhat Interface that I brought into common methods, this check included.